### ISSUE TITLE
Cargo timings graph is broken in Firefox

### ISSUE THREAD

**Issue by Lucretiel** [rust-lang/cargo#8850]
When I run `cargo +nightly -Z timings`, it produces a timings report in HTML as expected. This renders correctly in Chrome and Safari. However, in Firefox, I only see the summary table and timings breakdown table. I can see the controls for the graph, but the graph itself doesn't render: 
<img width="1247" alt="image" src="https://user-images.githubusercontent.com/1823405/98853598-5f87ec80-2427-11eb-94ea-af7366d7416e.png">

In the error console, I see the following uncaught exception:

```
Uncaught 
    Exception
        columnNumber: 0
        data: null
        filename: "http://localhost:8080/cargo-timing.html"
        lineNumber: 24173
        message: ""
        name: "NS_ERROR_FAILURE"
        result: 2147500037
        stack: "setup_canvas@http://localhost:8080/cargo-timing.html:24173:7\ndraw_graph_axes@http://localhost:8080/cargo-timing.html:24187:25\nrender_pipeline_graph@http://localhost:8080/cargo-timing.html:23930:86\n@http://localhost:8080/cargo-timing.html:24331:1\n"
        <prototype>: ExceptionPrototype { toString: toString(), name: Getter, message: Getter, … }
```

This exception is being thrown from the following function, specifically at `ctx.scale(dpr, dpr)` near the end:

```js
function setup_canvas(id, width, height) {
  let g = document.getElementById(id);
  let dpr = window.devicePixelRatio || 1;
  g.width = width * dpr;
  g.height = height * dpr;
  g.style.width = width;
  g.style.height = height;
  let ctx = g.getContext('2d');
  ctx.scale(dpr, dpr);  // Exception thrown here
  return ctx;
}
```

Digging into the debugger, there don't seem to be any obvious issue with that line of code. `ctx` is of prototype `CanvasRenderingContext2DPrototype`, and the `scale` method has been supported in firefox pretty much forever. `dpr == 2`.

### Meta

`rustc --version --verbose`:
```
cargo 1.49.0-nightly (79b397d72 2020-10-15)
release: 1.49.0
commit-hash: 79b397d72c557eb6444a2ba0dc00a211a226a35a
commit-date: 2020-10-15
```

- macOS Catalina, version 10.15.7 (19H2)
- Firefox 81.0.2 (64-bit)

</p>
</details>

**Comment by ehuss** [comment]
Transferred to the cargo repo.

Set the scale slider to the far left and reload the page, and then slowly raise the scale slider up to whatever limit it can handle.  The problem is that the canvas is too large.  AFAIK, there isn't a way to get the maximum canvas size (I think it may be graphics-card dependent), so it's kinda hard to know how to handle that.

**Comment by Lucretiel** [comment]
Thanks, will give it a try

**Comment by Lucretiel** [comment]
Yeah, that fixes it. Sliding the scale down doesn't do anything immediately, but refreshing afterwards does show the graphs.

**Comment by nagisa** [comment]
For me the problem seems to have to do with the height of the canvas being more than 32k pixels. The graph won’t render until I filter out enough crates with the "min duration" slider. Here's the relevant bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1282074

**Comment by bbb651** [comment]
I'm also in favor of a DOM/SVG based solution, it offers several immediate benefits:
- Making the text in the graph searchable
- Allowing it to be easily themed with a `prefers-color-scheme` media query for a simple dark theme or with browser extensions 
- Not requiring javascript

There are existing tools that use both approaches, for example [FlameGraph](https://github.com/brendangregg/FlameGraph) uses an SVG and [speedscope](https://github.com/jlfwong/speedscope) uses an interactive canvas interface.
I think making use of an existing tool, by either utilizing an existing benchmark file format or embedding the viewer it in the generated html would provide the best experience, but either way would probably be as much work as rewriting the entire thing...

**Comment by epage** [comment]
> Not requiring javascript

@eth3lbert your experiement looks to do all of the rendering through Javascript. Is that just for the specific phase of the experiment or your expected direction?  If the direction, could you talk to why you made that decision?

**Comment by eth3lbert** [comment]
> > Not requiring javascript
> 
> [@eth3lbert](https://github.com/eth3lbert) your experiement in [#15091](https://github.com/rust-lang/cargo/pull/15091) looks to do all of the rendering through Javascript. Is that just for the specific phase of the experiment or your expected direction? If the direction, could you talk to why you made that decision?

Good question! I first implemented this all in JS, which I think is easier when interaction comes into play and the rendering logic is similar to how it's rendered in canvas.

I considered output with the initial SVG rendered. But it would be significantly different from the current implementation. It might mostly involve manipulating class names to toggle visibility and the text of axes when considering interaction.
I haven't implemented it this way yet and I'm not sure how the performance difference will be. However, from my naive tests, the current approach provides acceptable performance. If others encounter severe performance issues, I could try iterating to new approaches.

**Comment by epage** [comment]
> However, from my naive tests, the current approach provides acceptable performance.

How big of packages?

> If others encounter severe performance issues, I could try iterating to new approaches.

This will be blocked until we feel confident on the performance of this.  If you'd like to move this forward, then it will be on you to build that confidence through evaluating different cases.

**Comment by eth3lbert** [comment]
> How big of packages?

I've tested this with `crates.io` which contains `627 units` in total. I'm not sure if this is enough, though!
Rendering large SVGs in Firefox can be a bit laggy while scrolling, but it's much better than having them completely broken :D


> This will be blocked until we feel confident on the performance of this. If you'd like to move this forward, then it will be on you to build that confidence through evaluating different cases.

No worries! I think we still want more testers and feedback anyway. I might also try other approaches later before feedback when I have more spare time.
